### PR TITLE
fix(components): [calendar] show wrong weekday when set weekStart

### LINKED ISSUE #9821: [Component] [calendar] `el-calendar` shows wrong weekday
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.39`
- Element Plus Version: `2.2.17`
- Browser / OS: `Edge 105.0.1343.42 / macOS 12.5`
- Build Tool: `Vite`

### Related Component

- `el-calendar`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8aDEgc3R5bGU9XCJjb2xvcjogcmVkO1wiPlxuICAgIFdyb25nOiAyMDE5LTA5LTAxIGlzIG5vdCBNb25kYXk7IGl0J3MgU3VuZGF5ISEhXG4gIDwvaDE+XG4gIDxlbC1jYWxlbmRhciB2LW1vZGVsPVwidmFsdWUyMDE5XCIgLz5cbiAgPGgxIHN0eWxlPVwiY29sb3I6IGdyZWVuO1wiPlxuICAgIENvcnJlY3Q6IDIwMjItMDktMDEgaXMgVGh1cnNkYXlcbiAgPC9oMT5cbiAgPGVsLWNhbGVuZGFyIHYtbW9kZWw9XCJ2YWx1ZTIwMjJcIiAvPlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdD5cbmltcG9ydCB7IGRheWpzIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuZGF5anMuZW4ud2Vla1N0YXJ0ID0gMVxuPC9zY3JpcHQ+XG5cbjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5jb25zdCB2YWx1ZTIwMTkgPSByZWYobmV3IERhdGUoJzIwMTktMDktMDEnKSlcbmNvbnN0IHZhbHVlMjAyMiA9IHJlZihuZXcgRGF0ZSgnMjAyMi0wOS0wMScpKVxuPC9zY3JpcHQ+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

1. Import the internal `dayjs` object from `element-plus`;
2. Set `dayjs.en.weekStart` to be `1` so that the calendar week starts at Monday, not Sunday;
3. Pass `new Date('2019-09-01')` as `v-model` to `el-calendar` component.

### What is Expected?

Sep 1, 2019 shows as Sunday, at the end of the first row in the calendar.

### What is actually happening?

Sep 1, 2019 shows as Monday, at the beginning of the first row in the calendar.

### Additional comments

It turns out that, whenever a month starts at Sunday, the `el-calendar` component will show its first day at the beginning of the first row, regardless of what value `dayjs.en.weekStart` is set to be.

Here is an incomplete list of such months (taken from the first 30 years of the 21st century): `2001-04`, `2001-07`, `2002-09`, `2002-12`, `2003-06`, `2004-02`, `2004-08`, `2005-05`, `2006-01`, `2006-10`, `2007-04`, `2007-07`, `2008-06`, `2009-02`, `2009-03`, `2009-11`, `2010-08`, `2011-05`, `2012-01`, `2012-04`, `2012-07`, `2013-09`, `2013-12`, `2014-06`, `2015-02`, `2015-03`, `2015-11`, `2016-05`, `2017-01`, `2017-10`, `2018-04`, `2018-07`, `2019-09`, `2019-12`, `2020-03`, `2020-11`, `2021-08`, `2022-05`, `2023-01`, `2023-10`, `2024-09`, `2024-12`, `2025-06`, `2026-02`, `2026-03`, `2026-11`, `2027-08`, `2028-10`, `2029-04`, `2029-07`, `2030-09`, `2030-12`.

Screenshot of the reproduction:
image

### PR DESCRIPTION
**Pull request by init-qy**
Please make sure these boxes are checked before submitting your PR, thank you!

closed #9821 

by the way,the unit test  `firstDayOfWeek in range mode` looks outdated,it still use prop `first-day-of-week`

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by init-qy** [comment]
[before](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8aDEgc3R5bGU9XCJjb2xvcjogcmVkO1wiPlxuICAgIFdyb25nOiAyMDE5LTA5LTAxIGlzIG5vdCBNb25kYXk7IGl0J3MgU3VuZGF5ISEhXG4gIDwvaDE+XG4gIDxlbC1jYWxlbmRhciB2LW1vZGVsPVwidmFsdWUyMDE5XCIgLz5cbiAgPGgxIHN0eWxlPVwiY29sb3I6IGdyZWVuO1wiPlxuICAgIENvcnJlY3Q6IDIwMjItMDktMDEgaXMgVGh1cnNkYXlcbiAgPC9oMT5cbiAgPGVsLWNhbGVuZGFyIHYtbW9kZWw9XCJ2YWx1ZTIwMjJcIiAvPlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdD5cbmltcG9ydCB7IGRheWpzIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuZGF5anMuZW4ud2Vla1N0YXJ0ID0gMVxuPC9zY3JpcHQ+XG5cbjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5jb25zdCB2YWx1ZTIwMTkgPSByZWYobmV3IERhdGUoJzIwMTktMDktMDEnKSlcbmNvbnN0IHZhbHVlMjAyMiA9IHJlZihuZXcgRGF0ZSgnMjAyMi0wOS0wMScpKVxuPC9zY3JpcHQ+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)
[after](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8aDEgc3R5bGU9XCJjb2xvcjogcmVkO1wiPlxuICAgIFdyb25nOiAyMDE5LTA5LTAxIGlzIG5vdCBNb25kYXk7IGl0J3MgU3VuZGF5ISEhXG4gIDwvaDE+XG4gIDxlbC1jYWxlbmRhciB2LW1vZGVsPVwidmFsdWUyMDE5XCIgLz5cbiAgPGgxIHN0eWxlPVwiY29sb3I6IGdyZWVuO1wiPlxuICAgIENvcnJlY3Q6IDIwMjItMDktMDEgaXMgVGh1cnNkYXlcbiAgPC9oMT5cbiAgPGVsLWNhbGVuZGFyIHYtbW9kZWw9XCJ2YWx1ZTIwMjJcIiAvPlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdD5cbmltcG9ydCB7IGRheWpzIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuZGF5anMuZW4ud2Vla1N0YXJ0ID0gMVxuPC9zY3JpcHQ+XG5cbjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5jb25zdCB2YWx1ZTIwMTkgPSByZWYobmV3IERhdGUoJzIwMTktMDktMDEnKSlcbmNvbnN0IHZhbHVlMjAyMiA9IHJlZihuZXcgRGF0ZSgnMjAyMi0wOS0wMScpKVxuPC9zY3JpcHQ+XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctOTgyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcInVuc3VwcG9ydGVkXCJcbiAgfVxufSIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9ydW50aW1lLWRvbUBsYXRlc3QvZGlzdC9ydW50aW1lLWRvbS5lc20tYnJvd3Nlci5qc1wiLFxuICAgIFwiQHZ1ZS9zaGFyZWRcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvc2hhcmVkQGxhdGVzdC9kaXN0L3NoYXJlZC5lc20tYnVuZGxlci5qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzXCI6IFwiaHR0cHM6Ly9wcmV2aWV3LTk4MjQtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJ1bnN1cHBvcnRlZFwiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcbmltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICByZXR1cm4gbmV3IFByb21pc2UoKHJlc29sdmUsIHJlamVjdCkgPT4ge1xuICAgIGNvbnN0IGxpbmsgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaW5rJylcbiAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgIGxpbmsuaHJlZiA9ICdodHRwczovL3ByZXZpZXctOTgyNC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmNzcydcbiAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2xvYWQnLCByZXNvbHZlKVxuICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignZXJyb3InLCByZWplY3QpXG4gICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgfSlcbn0iLCJfbyI6eyJzaG93SGlkZGVuIjp0cnVlLCJzdHlsZVNvdXJjZSI6Imh0dHBzOi8vcHJldmlldy05ODI0LWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguY3NzIn19)

**Review by tolking — APPROVED** [review]

**Review by holazz — APPROVED** [review]

**Comment by btea** [comment]
@init-qy Could you add a test case for this fix?

**Comment by init-qy** [comment]
> @init-qy Could you add a test case for this fix?

ok

File: packages/components/calendar/__tests__/calendar.test.tsx [hunk 1]
```diff
@@ -1,10 +1,20 @@
 import { nextTick } from 'vue'
 import { mount } from '@vue/test-utils'
 import { describe, expect, it } from 'vitest'
+import updateLocale from 'dayjs/plugin/updateLocale'
+import { dayjs } from '../../../element-plus/index'
```
Comment (btea) [inline packages/components/calendar/__tests__/calendar.test.tsx:5]:
```suggestion
import dayjs from 'dayjs'
```

**Review by tolking — APPROVED** [review]

**Pull request merged**